### PR TITLE
#12214: refactor moreh_sgd from deprecated to ttnn

### PR DESCRIPTION
**Pull request by DuongQLee**
### Ticket
[#12214](https://github.com/tenstorrent/tt-metal/issues/12214)

### Problem description
Refactor deprecated `moreh_sgd` in `tt_dnn` to `ttnn`.

### What's changed

- Remove old `moreh_sgd` implementation
- Move device kernels code to `ttnn`
- Create new wrapper code for `ttnn`

### Checklist
- [x] Post commit CI passes: [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/11140683689)
- [x] Blackhole Post commit (if applicable): Not available.
- [x] Model regression CI testing passes (if applicable): Not available.
- [x] New/Existing tests provide coverage for changes: 164/164 test cases passed.

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        ret.push_back(tensor_args.output_tensors.at(0).value());
+    } else {
+        ret.push_back(create_device_tensor(output_shapes.at(0).value(), dtype, layout, device, output_mem_config));
+    }
+
+    if (tensor_args.output_tensors.at(1).has_value()) {
+        ret.push_back(tensor_args.output_tensors.at(1).value());
+    } else {
+        ret.push_back(create_device_tensor(output_shapes.at(1).value(), dtype, layout, device, momentum_buffer_out_mem_config));
+    }
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp:80]:
```suggestion
    if (tensor_args.output_tensors.at(1).has_value()) {
        ret.push_back(tensor_args.output_tensors.at(1).value());
    } else if (this->momentum != 0.0f) {
        ret.push_back(create_device_tensor(output_shapes.at(1).value(), dtype, layout, device, momentum_buffer_out_mem_config));
    } else {
       ret.push_back(std::nullopt);
    }
```
The second output may not be necessary depending on the momentum value

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        {
+            {tt::CB::c_in0, 2},                      // param_in
+            {tt::CB::c_in1, 2},                      // grad
+            {tt::CB::c_in2, 2},                      // momentum_in
+            {tt::CB::c_out0, 2},                      // param_out
+            {tt::CB::c_out1, 2},                      // momentum_out
+
+            {tt::CB::c_intermed0, 5, fp32_dest_acc_en ? tt::DataFormat::Float32 : data_format},  // cb_scalar_args (lr, momentum, dampening, weight_decay, one)
+            {tt::CB::c_intermed1, 1, fp32_dest_acc_en ? tt::DataFormat::Float32 : data_format},  //
+            {tt::CB::c_intermed2, 1, fp32_dest_acc_en ? tt::DataFormat::Float32 : data_format},  //
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp:83]:
```suggestion
            {tt::CB::c_intermed1, 1, intermed_cb_format },  //
            {tt::CB::c_intermed2, 1, intermed_cb_format },  //
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& tensor_return_value) {
+    auto& program = cached_program.program;
+    auto& reader_kernel_id = cached_program.shared_variables.unary_reader_kernel_id;
+    auto& writer_kernel_id = cached_program.shared_variables.unary_writer_kernel_id;
+
+    auto param_in_buffer = tensor_args.param_in.buffer();
+    auto grad_buffer = tensor_args.grad.buffer();
+    // auto momentum_buffer_in_buffer = tensor_args.momentum_buffer_in->buffer();
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp:244]: Please remove the debug code

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        const std::optional<const Tensor> momentum_buffer_out,
+        float lr,
+        float momentum,
+        float dampening,
+        float weight_decay,
+        bool nesterov,
+        bool momentum_initialized,
+        const std::optional<MemoryConfig>& param_out_mem_config,
+        const std::optional<MemoryConfig>& momentum_buffer_out_mem_config,
+        const DeviceComputeKernelConfig compute_kernel_config);
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp:26]:
```suggestion
        const std::optional<DeviceComputeKernelConfig> compute_kernel_config);
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd_pybind.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+            py::arg("lr"),
+            py::arg("momentum"),
+            py::arg("dampening"),
+            py::arg("weight_decay"),
+            py::arg("nesterov"),
+            py::arg("momentum_initialized"),
+
+            py::arg("param_out_mem_config") = std::nullopt,
+            py::arg("momentum_buffer_out_mem_config") = std::nullopt,
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd_pybind.cpp:31]:
```suggestion
            py::arg("param_out_memory_config") = std::nullopt,
            py::arg("momentum_buffer_out_memory_config") = std::nullopt,
```

**Review by mrshaw01 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp (hunk 4 above)
Comment (mrshaw01) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp]: @hschoi4448 Is this resolved...

Comment (mrshaw01) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp]: @hschoi4448 I checked and it's resolved

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/decorators.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp:7]: Can we please avoid device operation include here? I don’t see anything from it used here.

Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp:7]: resolved in [285329d](https://github.com/tenstorrent/tt-metal/pull/12378/commits/285329d75aed17e5bd826e8ee69747bdc320c677)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        bool momentum_initialized,
+        const std::optional<MemoryConfig>& param_out_memory_config,
+        const std::optional<MemoryConfig>& momentum_buffer_out_memory_config,
+        const std::optional<DeviceComputeKernelConfig>& compute_kernel_config);
+};
+}  // namespace ttnn::operations::moreh::moreh_sgd
+
+namespace ttnn {
+constexpr auto moreh_sgd =
+    ttnn::register_operation<"ttnn::moreh_sgd", ttnn::operations::moreh::moreh_sgd::MorehSgd>();
```
[the remark this answers is not in the document]
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp]: resolved in [b639c02](https://github.com/tenstorrent/tt-metal/pull/12378/commits/b639c029616704f71974e7b449d677c608266feb)

File: ttnn/CMakeLists.txt
Comment (mrshaw01) [inline ttnn/CMakeLists.txt:1]: Please do sorting (if not yet) and remove the space.

File: ttnn/cpp/ttnn/operations/moreh/moreh_pybind.cpp
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_pybind.cpp:1]: Please sort lines.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    if (tensor_args.output_tensors.at(1).has_value()) {
+        tt::operations::primary::check_tensor(
+            tensor_args.output_tensors.at(1).value(), "moreh_sgd", "momentum_buffer_out");
+    }
+}
+
+MorehSgdOperation::program_factory_t MorehSgdOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    // For now we litteraly don't care and return a single factory. Whatever
+    return ProgramFactory{};
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp:40]: Remove comment

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        const std::optional<const Tensor> momentum_buffer_out,
+        float lr,
+        float momentum,
+        float dampening,
+        float weight_decay,
+        bool nesterov,
+        bool momentum_initialized,
+        const std::optional<MemoryConfig>& param_out_memory_config,
+        const std::optional<MemoryConfig>& momentum_buffer_out_memory_config,
+        const DeviceComputeKernelConfig compute_kernel_config);
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp:26]: The concrete compute_kernel_config here seems incorrect. It should be std::optional throughout until the primitive is invoked.

**Review by mrshaw01 — CHANGES_REQUESTED** [review]

**Comment by mrshaw01** [comment]
Please revert to: https://github.com/tenstorrent/tt-metal/commit/85b6ec1c842f1ce0cfa8f9e9d7ea29f994a16920 and address Artem's comment.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp (hunk 9 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp]: resolved in [dc2ce28](https://github.com/tenstorrent/tt-metal/pull/12378/commits/dc2ce28ff17ddff2b33f87920282e226d4fec218)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp (hunk 8 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp]: resolved in [dc2ce28](https://github.com/tenstorrent/tt-metal/pull/12378/commits/dc2ce28ff17ddff2b33f87920282e226d4fec218)

File: ttnn/cpp/ttnn/operations/moreh/moreh_pybind.cpp
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_pybind.cpp:1]: resolved in [dc2ce28](https://github.com/tenstorrent/tt-metal/pull/12378/commits/dc2ce28ff17ddff2b33f87920282e226d4fec218)

File: ttnn/CMakeLists.txt
Comment (DuongQLee) [reply ttnn/CMakeLists.txt:1]: resolved in [dc2ce28](https://github.com/tenstorrent/tt-metal/pull/12378/commits/dc2ce28ff17ddff2b33f87920282e226d4fec218)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    static void validate_on_program_cache_miss(const operation_attributes_t&, const tensor_args_t&);
+    static void validate_on_program_cache_hit(const operation_attributes_t&, const tensor_args_t&);
+    static shape_return_value_t compute_output_shapes(const operation_attributes_t&, const tensor_args_t&);
+    static tensor_return_value_t create_output_tensors(const operation_attributes_t&, const tensor_args_t&);
+    static std::tuple<operation_attributes_t, tensor_args_t> invoke(
+        const Tensor& param_in,
+        const Tensor& grad,
+        const std::optional<const Tensor> momentum_buffer_in,
+        const std::optional<const Tensor> param_out,
+        const std::optional<const Tensor> momentum_buffer_out,
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp:70]: Use `const std::optional<Tensor>&`

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+
+namespace ttnn::operations::moreh::moreh_sgd {
+struct MorehSgd {
+    static std::vector<std::optional<Tensor>> invoke(
+        const Tensor& param_in,
+        const Tensor& grad,
+        const std::optional<const Tensor> momentum_buffer_in,
+        const std::optional<const Tensor> param_out,
+        const std::optional<const Tensor> momentum_buffer_out,
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp:17]: Use const std::optional<Tensor>&

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        bool momentum_initialized;
+        const MemoryConfig param_out_memory_config;
+        const MemoryConfig momentum_buffer_out_memory_config;
+        const DeviceComputeKernelConfig compute_kernel_config;
+    };
+
+    struct tensor_args_t {
+        const Tensor& param_in;
+        const Tensor& grad;
+        const std::optional<const Tensor> momentum_buffer_in;
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp:27]: Use const std::optional<Tensor>&

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+            operation_attributes.momentum_buffer_out_memory_config));
+    }
+
+    return std::move(ret);
+}
+
+std::tuple<MorehSgdOperation::operation_attributes_t, MorehSgdOperation::tensor_args_t> MorehSgdOperation::invoke(
+    const Tensor& param_in,
+    const Tensor& grad,
+    const std::optional<const Tensor> momentum_buffer_in,
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp:90]: Use const std::optional<Tensor>&

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "moreh_sgd.hpp"
+
+#include "ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp"
+
+namespace ttnn::operations::moreh::moreh_sgd {
+std::vector<std::optional<Tensor>> MorehSgd::invoke(
+    const Tensor& param_in,
+    const Tensor& grad,
+    const std::optional<const Tensor> momentum_buffer_in,
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp:13]: Use const std::optional<Tensor>&

Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp]: resolved in [a981edd](https://github.com/tenstorrent/tt-metal/pull/12378/commits/a981edd24c59337a422bc4c59ae88b14489c75a5)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp (hunk 12 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp:27]: resolved in [a981edd](https://github.com/tenstorrent/tt-metal/pull/12378/commits/a981edd24c59337a422bc4c59ae88b14489c75a5)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp (hunk 11 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.hpp]: resolved in [a981edd](https://github.com/tenstorrent/tt-metal/pull/12378/commits/a981edd24c59337a422bc4c59ae88b14489c75a5)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp (hunk 10 above)
Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.hpp]: resolved in [a981edd](https://github.com/tenstorrent/tt-metal/pull/12378/commits/a981edd24c59337a422bc4c59ae88b14489c75a5)

**Review by mrshaw01 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    float lr,
+    float momentum,
+    float dampening,
+    float weight_decay,
+    bool nesterov,
+    bool momentum_initialized,
+    const std::optional<MemoryConfig>& param_out_memory_config,
+    const std::optional<MemoryConfig>& momentum_buffer_out_memory_config,
+    const std::optional<DeviceComputeKernelConfig>& compute_kernel_config) {
+    return {true, true};
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp:66]: why is return type vector<optional<Tensor>> if this method unconditionally creates both outputs?

**Review by ayerofieiev-tt — APPROVED** [review]
Overall lgtm.

Please review the return type here with the team, something is off right now.

Leaving approve to let you move faster once you resolve the above with the team.
Thank you

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    if (tensor_args.momentum_buffer_out.has_value()) {
+        ret.push_back(tensor_args.momentum_buffer_out.value());
+    } else {
+        ret.push_back(create_device_tensor(
+            output_shapes.at(1).value(),
+            dtype,
+            layout,
+            device,
+            operation_attributes.momentum_buffer_out_memory_config));
+    }
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp:80]: `momentum_buffer_out` only produces output when the momentum is not zero.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp (hunk 15 above)
Comment (hschoi4448) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/moreh_sgd.cpp:66]:
```suggestion
    return {true, momentum != 0};
```
The first parameter is always t`rue`, but the second parameter, `momentum_buffer_out`, should be determined based on the value of the `momentum`.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp (hunk 16 above)
Comment (mrshaw01) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp:80]: I resolved this once, but Duong mistakenly force-pushed and removed the changes. When you make changes in `device_operation.cpp`, you also need to update `program_factory.cpp`
@DuongQLee 

Reference: https://github.com/tenstorrent/tt-metal/blob/85b6ec1c842f1ce0cfa8f9e9d7ea29f994a16920/ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_program_factory.cpp

Comment (DuongQLee) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp]: I have resolved this.

**Comment by DuongQLee** [comment]
> Overall lgtm.
> 
> Please review the return type here with the team, something is off right now.
> 
> Leaving approve to let you move faster once you resolve the above with the team. Thank you

I have review the return type, and edit the create asycn flag. This operation still needs to return std::vector<optional<Tensor>>.
Thank you.

**Pull request merged**